src/charmed_kubeflow_chisme/kubernetes/_kubernetes_resource_handler.py

```diff
@@ -289,3 +412,95 @@ def _charm_status_given_resource_status(
+    }
+
+
+def _get_resource_classes_in_manifests(resource_list: LightkubeResourcesList):
+    """Returns a list of the resource classes in a list of resources."""
+    resource_classes = []
+    for resource in resource_list:
+        if type(resource) not in resource_classes:
+            resource_classes.append(type(resource))
+    return resource_classes
+
+
+def _hash_lightkube_resource(resource: Resource) -> Tuple[str, str, str, str, str]:
+    """Hashes a Lightkube Resource by returning a tuple of (group, version, kind, name, namespace).
+
+    For global resources or resources without a namespace specified, namespace will be None.
+    """
+    resource_info = api_info(resource).resource
+
+    return (
+        resource_info.group,
+        resource_info.version,
+        resource_info.kind,
+        resource.metadata.name,
+        resource.metadata.namespace,
+    )
+
+
+def _in_left_not_right(left: list, right: list, hasher: Optional[Callable] = None) -> list:
+    """Returns the items in left that are not right (the Set difference).
+
+    Args:
+        left: a list
+        right: a list
+        hasher: (Optional) a function that hashes the items in left and right to something
+                immutable that can be compared.  If omitted, will use hash()
+
+    Returns:
+        A list of items in left that are not in right, based on the hasher function.
+    """
+    if hasher is None:
+        hasher = hash
+
+    left_as_dict = {hasher(resource): resource for resource in left}
+    right_as_dict = {hasher(resource): resource for resource in right}
+
+    keys_in_left_not_right = set(left_as_dict.keys()) - set(right_as_dict.keys())
+    items_in_left_not_right = [left_as_dict[k] for k in keys_in_left_not_right]
+
+    return items_in_left_not_right
+
+
+def _validate_labels_and_child_resource_types(labels, child_resource_types, caller_name):
+    """Validates labels and child_resource_types, raising a ValueError if either is empty."""
+    if not labels:
+        raise ValueError(ERROR_MESSAGE_NO_LABELS.format(caller=caller_name))
+    if not child_resource_types:
+        raise ValueError(ERROR_MESSAGE_NO_CHILD_RESOURCE_TYPES.format(caller=caller_name))
+
+
+def _validate_resources(resources, allowed_resource_types: LightkubeResourceTypesList):
+    """Validates that the resources are of a type in the allowed_resource_types list.
+
+    Side effect: raises a ValueError if any resource is not in the allowed_resource_types list.
+
+    Args:
+        resources: a list of Lightkube resources to validate
+        allowed_resource_types: a list of Lightkube resource classes to validate against
+    """
+    resource_types = _get_resource_classes_in_manifests(resources)
+    for resource_type in resource_types:
+        if resource_type not in allowed_resource_types:
+            raise ValueError(
+                f"Resource type {resource_type} not in allowed resource types"
+                f" '{allowed_resource_types}'"
```

Comment:
Can we also have it in logs?

Reply:
Should this helper write to a logger, or should the charm calling this log the error? I'm never sure where to put these things.

If we put a log statement here, then it'll always log the error even if whoever called this function (eg: a charm) caught it and handled it. Would that be confusing?